rfc/0004-at_plus.md

```diff
@@ -0,0 +1,207 @@
+
+# General Design
+In the big picture of `AT_PLUS` for the developer, there are two components that
+must be considered:
+
+ 1. Mutations are executed on a bucket. They might or might not include
+    `MutationTokens` as part of the response.
+ 2. The application state after the mutation happened needs to be applied at
+    N1QL query time.
+
+This RFC does not focus on 1. since it has been implemented already as part of
+the enhanced durability requirements initiative in the 4.0 timeframe. This RFC
+assumes that the Mutation Token is (at least) available as a return value from
+a mutation operation, most likely exposed on the `Document` or equivalent or 
+stored internally (see implementation details). `AT_PLUS` is not possible without
+`MutationTokens`.
+
+Note that this RFC proposes an extension of the current `MutationToken` to also
+include the bucket `name` as a reference. This is needed to prevent the user from
+mistakenly pass a mutation token from one bucket into another one (and to resolve
+issues around multi-bucket n1ql queries).
+
+Since it is important to this RFC, the properties of a `MutationToken` are as
+follows:
+
+ 1. A `MutationToken` is opaque to the user and needs to be treated as such.
+ 2. A `MutationToken` is immutable after creation.
+
+The internal structure of a Mutation Token is a triple which contains:
+
+```
+MutationToken {
+    [required] vbucket_id = long
+    [required] vbucket_uuid = long
+    [required] sequence_number = long
+    [required] bucket = string/reference (either name of the bucket or immutable "reference" to the target bucket)
+}
+```
+
+The N1QL query execution for AT_PLUS consists of the following user level API:
+
+```
+consistent_with(bucket: Bucket, [docs: Document...])
+```
+
+This method must have the following behavior:
+
+ - When called with just the `Bucket` reference and no `Documents`, this means that
+   the N1QL query must be consistent with all mutations that happened against this
+   `Bucket`.
+ - When called with `Bucket` and a list of `Documents` and the `MutationTokens` are
+   present on the `Document`, then they are extracted and used as part of the query.
+   This is the most performant option in terms of network overhead.
+ - When When called with `Bucket` and a list of `Documents` and the `MutationTokens` is
+   not present on a specific `Document`, the ID is used, hashed and the specific token
+   for that vbucket is used.
+ - Subsequent calls to either a previously captured `Document` on the same `Bucket` 
+   reference will override the old one, there must not be duplicates in the performed
+   query against N1QL.
+
+Setting `SCAN_CONSISTENCY` to anything but the default is considered a runtime
+error (see errors). This means that `AT_PLUS`, while needed internally, is never
+actually exposed to the user level. The user only specifies the consistentWith
+boundaries and the SDK needs to figure out the rest (see implementation).
+
```

Comment:
@jeffrymorris yes, the idea was that setting something on `consistent_with` implies "at_plus", so there is no need to set the param explicitly (which means we also need to export it on the enum).